compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1033,6 +1037,11 @@ private boolean isTupleTypeAssignable(BType source, BType target, Set<TypePair>
             unresolvedTypes.add(pair);
         }
 
```

Comment:
:recycle: Think we cache the result for `source.tag == TypeTags.TUPLE` since it is used in many places in this method. We can also populate `rhsTupleType` beforehand. 

:thinking: :question: I am wondering why the `tupleTypes` has to be empty for this scenario.

Reply:
Please revisit point 2. Logic needs changes w.r.t. to JSON type. Probably need more test cases too.